File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/C-Кузнечик_собирает_монеты/solution.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <climits>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n, k;
    std::scanf("%d %d", &n, &k);
    std::vector<long long> dp(n + 1, LLONG_MAX);
    dp[1] = 0;
    std::vector<int> nums = {0, 0};
    for (int i = 0; i < n - 2; ++i) {
        int num;
        std::scanf("%d", &num);
        nums.push_back(num);
    }
    nums.push_back(0);
    std::deque<int> q;
    q.push_back(1);
    std::vector<int> prev (n + 1, -1);
    for (int i = 2; i <= n; ++i) {
        if (!q.empty() && q.front() < i - k) q.pop_front();
        dp[i] = dp[q.front()] + nums[i];
        prev[i] = q.front();
        while (!q.empty() && dp[q.back()] <= dp[i]) q.pop_back();
        q.push_back(i);
    }
    std::vector<int> path;
    int i = n;
    while (prev[i] != -1) {
        path.push_back(prev[i]);
        i = prev[i];
    }
    
    std::printf("%lld\n%d\n", dp[n], path.size());
    for (int j = path.size() - 1; j >= 0; --j) std::printf("%d ", path[j]);
    std::printf("%d\n", n);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/C-Кузнечик_собирает_монеты/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n, k = map(int, sys.stdin.readline().rstrip().split())
    nums: list[int] = [0, 0] + list(map(int, sys.stdin.readline().rstrip().split()))
    nums.append(0)
    dp: list[int] = [float('inf')] * (n + 1)
    dp[1] = 0
    q: list[int] = deque()
    q.append(1)
    prev: list[int] = [-1] * (n + 1)
    for i in range(2, n + 1):
        if q and q[0] < i - k: q.popleft()
        dp[i] = dp[q[0]] + nums[i]
        prev[i] = q[0]
        while q and dp[q[-1]] <= dp[i]: q.pop()
        q.append(i)
    path: list[int] = [n]
    i: int = n
    while prev[i] != -1:
        path.append(prev[i])
        i = prev[i]
    sys.stdout.write('{}\n{}\n{}\n'.format(dp[n], len(path) - 1, ' '.join(str(move) for move in reversed(path))))


if __name__ == '__main__':
    solution()